servicetalk-opentracing-log4j2/src/main/java/io/servicetalk/opentracing/log4j2/ServiceTalkTracingThreadContextMap.java

```diff
@@ -132,116 +114,34 @@ public boolean isEmpty() {
-            public Map<String, String> toMap() {
-                return getCopy(storage, SCOPE_MANAGER.active());
-            }
-
-            @Override
-            public boolean containsKey(String key) {
-                return containsTracingKey(key) || storage.containsKey(key);
-            }
-
-            @SuppressWarnings("unchecked")
-            @Override
-            public <V> void forEach(BiConsumer<String, ? super V> action) {
-                InMemoryScope scope = SCOPE_MANAGER.active();
-                if (scope != null) {
-                    InMemorySpan span = scope.span();
-                    action.accept(TRACE_ID_KEY, (V) span.traceIdHex());
-                    action.accept(SPAN_ID_KEY, (V) span.spanIdHex());
-                    action.accept(PARENT_SPAN_ID_KEY, (V) span.nonnullParentSpanIdHex());
-                }
-                storage.forEach((key, value) -> action.accept(key, (V) value));
-            }
-
-            @SuppressWarnings("unchecked")
-            @Override
-            public <V, S> void forEach(TriConsumer<String, ? super V, S> action, S state) {
-                InMemoryScope scope = SCOPE_MANAGER.active();
-                if (scope != null) {
-                    InMemorySpan span = scope.span();
-                    action.accept(TRACE_ID_KEY, (V) span.traceIdHex(), state);
-                    action.accept(SPAN_ID_KEY, (V) span.spanIdHex(), state);
-                    action.accept(PARENT_SPAN_ID_KEY, (V) span.nonnullParentSpanIdHex(), state);
-                }
-                storage.forEach((key, value) -> action.accept(key, (V) value, state));
-            }
-
-            @SuppressWarnings("unchecked")
-            @Override
-            public <V> V getValue(String key) {
-                switch (key) {
-                    case TRACE_ID_KEY: {
-                        InMemoryScope scope = SCOPE_MANAGER.active();
-                        if (scope != null) {
-                            return (V) scope.span().traceIdHex();
-                        }
-                        break;
-                    }
-                    case SPAN_ID_KEY: {
-                        InMemoryScope scope = SCOPE_MANAGER.active();
-                        if (scope != null) {
-                            return (V) scope.span().spanIdHex();
-                        }
-                        break;
-                    }
-                    case PARENT_SPAN_ID_KEY: {
-                        InMemoryScope scope = SCOPE_MANAGER.active();
-                        if (scope != null) {
-                            return (V) scope.span().nonnullParentSpanIdHex();
-                        }
-                        break;
-                    }
-                    default:
-                        break;
-                }
-                return (V) storage.get(key);
-            }
+    public StringMap getReadOnlyContextData() {
+        StringMap map = new JdkMapAdapterStringMap(getCopy());
```

Comment:
@NiteshKant - note this is an API breaking change. The previous APIs leaked the underlying `Map` to avoid copy operations as an optimization. Now that the `Map` contains "token" elements that need to be unwrapped it leaks more knowledge and would require exposing the wrap/unwrap of `null`. I don't expect folks will be extending/calling these APIs so the risk is low, and it reduces our API footprint, at the cost of some breakage and less optimal implementation.